### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by Xiscoder**

### CODE REVIEW

File: src/Cql.InventoryManagement.DbMigrations/Migrations/2019/07/20190711114010-adding-plugins-datebought.cs [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    [Migration(20190711114010)]
+    public class _20190711114010_adding_plugins_datebought : AutoReversingMigration
+    {
+        public override void Up()
+        {
+            Alter.Table("Plugins")
+                .AddColumn("DateBought")
+                .AsDateTime()
+                .NotNullable()
+                .WithDefaultValue(1800);
```
Comment (freakingawesome) [inline src/Cql.InventoryManagement.DbMigrations/Migrations/2019/07/20190711114010-adding-plugins-datebought.cs]: Why the hardcoded date and not null? Leaving this column nullable seems more intuitive and less likely to cause problems down the road.

File: src/backend-api/Controllers/UpdateController.cs [hunk 2]
```diff
@@ -94,17 +90,43 @@ public IActionResult EditPrograms([FromBody] UpdateProgramInputModel input)
             {
                 foreach (var program in _context.Program.Where(x => x.ProgramName == input.Program.OldProgramName))
```
Comment (freakingawesome) [inline src/backend-api/Controllers/UpdateController.cs]: Is looking up by OldProgramName guaranteed be isolated to the program(s) desired? Is there a unique constraint on program names? Perhaps an ID lookup would be better here?

(I realize this isn't part of the changeset but figured I'd ask anyways)

File: src/backend-api/Controllers/UpdateController.cs [hunk 3]
```diff
@@ -229,6 +253,196 @@ public IActionResult EditProgram([FromBody] PostProgramInputModel input, [FromRo
...
+                    if (input.HardwareAssigned != null)
+                    {
+                        // loop through hardware and depending on what type the hardware is, then add the hardware to the specific table. 
+                        // also in the loop, update the hardware history whether the it was assigned or unassigned  
+                        foreach (var hardware in input.HardwareAssigned)
+                        {
+                            switch (hardware.Type.ToLower())
+                            {
+                                case "monitor":
+                                    var mon = _context.Monitor.Find(hardware.ID);
```
Comment (freakingawesome) [inline src/backend-api/Controllers/UpdateController.cs]: There's a lot of duplication here. Perhaps this is a good candidate for something similar to the IAssignable and ISoftDeletable interfaces? (or at least I think that's what they're called)

File: src/backend-api/Controllers/UpdateController.cs [hunk 4]
```diff
@@ -229,6 +253,196 @@ public IActionResult EditProgram([FromBody] PostProgramInputModel input, [FromRo
...
+                    }
+                    // checking if any hardware is to be unassigned
+                    if (input.HardwareUnassigned != null)
+                    {
+                        foreach (var hardware in input.HardwareUnassigned)
+                        {
+                            switch (hardware.Type.ToLower())
+                            {
+                                case "monitor":
+                                    var mon = _context.Monitor.Find(hardware.ID);
```
Comment (freakingawesome) [inline src/backend-api/Controllers/UpdateController.cs]: Same here - lots of duplication

File: src/backend-api/Controllers/UpdateController.cs (hunk 2 above)
Comment (Xiscoder) [reply src/backend-api/Controllers/UpdateController.cs]: Yes this does exactly as desired. We are seeking to edit all the programs here with a certain name.

File: src/Cql.InventoryManagement.DbMigrations/Migrations/2019/07/20190711114010-adding-plugins-datebought.cs (hunk 1 above)
Comment (Xiscoder) [reply src/Cql.InventoryManagement.DbMigrations/Migrations/2019/07/20190711114010-adding-plugins-datebought.cs]: Hmm, we could leave it nullable but there should always be a dateBought with every plugin. I'll change it to nullable.

File: src/backend-api/Controllers/UpdateController.cs (hunk 3 above)
Comment (Xiscoder) [reply src/backend-api/Controllers/UpdateController.cs]: Yup, I agree. I'll see what I can do

**Pull request merged**